google/cloud/status_or.h

```diff
@@ -24,33 +24,54 @@ namespace google {
 namespace cloud {
 inline namespace GOOGLE_CLOUD_CPP_NS {
 /**
- * `StatusOr<T>` is used to return a value or an error status.
+ * Holds a value or a `Status` indicating why there is no value.
  *
- * If the library is compiled with exceptions disabled there is no standard C++
- * mechanism to report errors. In that case we use this class to wrap the values
- * returned to the application.
- *
- * The application typically calls an API in the library and must check the
- * returned error status:
+ * `StatusOr<T>` represents either a usable `T` value or a `Status` object
+ * explaining why a `T` value is not present. Typical usage of `StatusOr<T>`
+ * looks like usage of a smart pointer, or even a std::optional<T>, in that you
+ * first check its validity using a conversion to bool (or by calling
+ * `StatusOr<T>::ok()`), then you may dereference the object to access the
+ * contained value. It is undefined behavior (UB) to dereference a
+ * `StatusOr<T>` that is not "ok". For example:
  *
  * @code
- * namespace gcs = google::cloud::storage;
- * void AppCode(gcs::noex::Client client) {
- *   gcs::StatusOr<gcs::BucketMetadata> meta_err = client.GetBucketMetadata(
- *       "my-bucket-name");
- *   if (!meta_err) {
- *       std::cerr << "Error in GetBucketMetadata: " << meta_err.status()
- *                 << std::endl;
- *       return;
- *   }
- *   gcs::BucketMetadata meta = *meta_err;
- *   // Do useful work here.
+ * StatusOr<Foo> foo = FetchFoo();
+ * if (!foo) {  // Same as !foo.ok()
+ *   // handle error and probably look at foo.status()
  * }
+ * foo->DoSomethingFooey();  // UB if !foo
```

Comment:
Perhaps putting this inside an "else" is clearer when the "handle error" code lacks an explicit return.

Reply:
Done.